src/main/java/com/marklogic/spark/writer/file/ZipFileWriter.java

```diff
@@ -57,8 +57,8 @@ public void write(InternalRow row) throws IOException {
         zipOutputStream.putNextEntry(new ZipEntry(entryName));
         this.contentWriter.writeContent(row, zipOutputStream);
         zipEntryCounter++;
-        if(properties.get(Options.READ_DOCUMENTS_CATEGORIES) != null &&
-            properties.get(Options.READ_DOCUMENTS_CATEGORIES).contains("metadata")){
+        if(!row.isNullAt(3) || (properties.get(Options.READ_DOCUMENTS_CATEGORIES) != null &&
```

Comment:
This should ignore the properties - just because they exist doesn't mean the rows have any metadata (also, just checking for "metadata" wouldn't suffice - the user could do "content,collections"). We instead want:

```
if (!row.isNullAt(3) || !row.isNullAt(4) || !row.isNullAt(5) || 
    !row.isNullAt(6) || !row.isNullAt(7)) {
    // add metadata entry
}
```

That also means that we'd want to test each of those conditions. So you should have 5 separate tests - one where collections is populated, then permissions, then quality, then properties, then metadata values. Those tests don't necessarily need to examine the ".metadata" file, they just need to verify that the ".metadata" file exists.

Reply:
User can have Options.READ_DOCUMENTS_CATEGORIES = "content,collections" as well right.

Reply:
Right, but the options aren't reliable as an indicator. We'll use the `READ_DOCUMENTS_CATEGORIES` option in our tests to make life easy. But a user could manually construct a Spark dataframe and populate the metadata column manually, in which case the `READ_DOCUMENTS_CATEGORIES` option won't exist. 

This is all due to the reader and writer being decoupled in Spark. The writer can't assume anything about how the data was read. So it can't assume that `READ_DOCUMENTS_CATEGORIES` was used. All it can do is look at the columns in the row.